Document/0x05c-Reverse-Engineering-and-Tampering.md

```diff
@@ -290,6 +290,91 @@ For white-box source code testing, you'll need a setup similar to the developer'
+
+###### Java
+
+As we already know, all the Java bytecode of an Android application is compiled into a dex file. Each dex file contains a [string identifiers list](https://source.android.com/devices/tech/dalvik/dex-format#file-layout "string identifiers list") (strings_ids). Each string used in the code is given an identifier number, and internally this identifier number is used whenever a string is referred. This list contains all the string identifiers used in the binary, including internal naming (e.g, type descriptors) or constant objects referred by the code (e.g hard coded strings). Knowing this, we can simply dump this list to get all the strings used in the Java bytecode in a dex file. Fortunately, there are plenty of tools already available for dumping strings form Java bytecode and one can choose depending on the needs. In this section we will use Ghidra (GUI based), and [Dextra](http://newandroidbook.com/tools/dextra.html "Dextra") (CLI based).
+
+With Ghidra, strings can be obtained by simply loading the dex file into the tool and by selecting **Window -> Defined strings** in the menu.
+
+> Loading APK file directly to Ghidra is buggy, thus it is recommended to extract dex file by unzipping the APK file and using it in Ghidra.
+
+<img src="Images/Chapters/0x05c/ghidra_dex_strings.png" width="500px"/>
+
+
+Dextra is a CLI based tool, and provides multiple features to explore dex files. All the strings in a dex file can be dumped using Dextra using the following command:
+
+```bash
+dextra -S classes.dex
+```
+
+The output from Dextra can be manipulated using standard Linux commands, for example, using `grep` to search for certain keywords.
+
+It is important to know, the list of strings obtained using the above tools can be very big, as it also includes the various class and package names used in the application. Going through the complete list, specially for big binaries, can be very cumbersome. Thus, it is recommended to start with keyword based searching and go through the list only when keyword search does not work. Some keywords which can be a good starting point are - password, key, and secret.
+
+
+###### Native
+
+For extracting strings from the native code used in an Android application, we can either use Ghidra (GUI based) or `strings` utility command available on \*nix.
+
+Use of Ghidra for extracting strings for a native binary has already been discussed in "[Using Disassemblers and Decompilers](0x04c-tampering-and-reverse-engineering#ghidra "Using Disassembler and Decompilers")" section.
+
+`strings` is a CLI based tool and using it is straightforward, as demonstrated below for a native library:
```

Comment:
```suggestion
`strings` is a CLI based tool and is pretty straightforward to use, as demonstrated below for a native library:
```

Reply:
You could also add to pipe the output of the strings command into a text file (strings a.so > output.txt), so it's easier to go through the output with other unix tools, like grep, head etc.

Reply:
Updated the content.